Device List when no controller found and enumeration optimisations

enodatio opened the pull request:
I made some changes to give better feedback when devices/ini files are not found and to make it clear to me whether it was directx not enumerating the device or altinput not finding a config match. Also a test to see if the ini file is found.

A few internal members were needed as flight gives feedback to the user through a screen message but the information is collected by config.

Also reordered the config parseinputs() loops to iterate on directinput.GetDevices(..) rather than looping the config input1... inputx  as this allows earlier termination in the event of a match and doesn't scan the file 128 times if there are no connected controllers.  Should be much more efficient in most common cases and makes collecting the list of connected controllers (FoundDevices) cleaner.

enodatio committed:
Add meaningful message if config file is not present.  (previously states version 0.0 does not match)

enodatio committed:
Reordered to query directinput.GetDevices once as more likely to be one joystick.  Added early termination to iterations on NumDevices if matching config is found.   Changed message to say no configured controller rather than no controller.  Added device list that shows the found Direct input joysticks to the debug message if no configured stick is found.

enodatio committed:
Added list of found controllers to screen message where no valid controller is found.  Made a few static internal members to facilitate this.

pbatard commented:
Shoot, I've been doing pretty much the exact same thing (listing all controllers, more explicit message if config is not present), and I have had these changes lined up since yesterday as well (haven't pushed them, as I need to clean them up a bit).

So we've really been duplicating work here... :(

Fortunately, there's some other stuff you picked that I don't have yet, so I guess what I'll do is push my changes, and then go over your proposal and pick whatever is left that I don't have.

At any rate, thank you very much for volunteering a patch!

enodatio commented:
Well its only about 10 lines and some cut and paste, but it's of use.  I wanted better info for debugging the condition where it doesn't load.  Made sense to offer it as a solution to the common complaints in the forum thread,

enodatio committed:
Seems to be a fix for the B9/Many loaded parts bug.
Made the Config Start Instantly and made the dll path relative so it can go in GameData

pbatard commented:
Happy to hear you didn't spend to much time.
I have now of run away with your idea of optimizing the device enumeration, and decided it's probably a lot smarter to drop the

```
[input#]
Name = Game Controller
```

to go with a much simpler

```
[Game Controller]
```

This should effectively remove the need to a max number of controller, and speed up the lookup.
I have now applied this change in 210cbf131c6d26da4512b31e8af9e62e31709257, and I hope with this last commit, I have more or less covered the improvements we've all been wanting.
